src/storage/src/hummock/event_handler/refiller.rs

```diff
@@ -0,0 +1,314 @@
+pub struct CacheRefiller {
+    /// order: old => new
+    queue: VecDeque<Item>,
+
+    context: CacheRefillContext,
+}
+
+impl CacheRefiller {
+    pub fn new(config: CacheRefillConfig, sstable_store: SstableStoreRef) -> Self {
+        let config = Arc::new(config);
+        let concurrency = Arc::new(Semaphore::new(config.concurrency));
+        Self {
+            queue: VecDeque::new(),
+            context: CacheRefillContext {
+                config,
+                concurrency,
+                sstable_store,
+            },
+        }
+    }
+
+    pub fn start_cache_refill(
+        &mut self,
+        deltas: Vec<SstDeltaInfo>,
+        pinned_version: Arc<PinnedVersion>,
+        new_pinned_version: PinnedVersion,
+    ) {
+        let task = CacheRefillTask {
+            deltas,
+            context: self.context.clone(),
+        };
+        let event = CacheRefillerEvent {
+            pinned_version,
+            new_pinned_version,
+        };
+        let handle = tokio::spawn(task.run());
+        let item = Item { handle, event };
+        self.queue.push_back(item);
+        GLOBAL_CACHE_REFILL_METRICS.refill_queue_total.add(1);
+    }
+
+    pub fn last_new_pinned_version(&self) -> Option<&PinnedVersion> {
+        self.queue.back().map(|item| &item.event.new_pinned_version)
+    }
+
+    pub fn next_event(&mut self) -> NextCacheRefillerEvent<'_> {
+        NextCacheRefillerEvent { refiller: self }
+    }
+}
+
+pub struct NextCacheRefillerEvent<'a> {
+    refiller: &'a mut CacheRefiller,
+}
+
+impl<'a> Future for NextCacheRefillerEvent<'a> {
+    type Output = CacheRefillerEvent;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let refiller = &mut self.deref_mut().refiller;
+
+        if let Some(item) = refiller.queue.front_mut() {
+            ready!(item.handle.poll_unpin(cx)).unwrap();
+            let item = refiller.queue.pop_front().unwrap();
+            GLOBAL_CACHE_REFILL_METRICS.refill_queue_total.sub(1);
+            return Poll::Ready(item.event);
+        }
+        Poll::Pending
+    }
+}
+
+pub struct CacheRefillerEvent {
+    pub pinned_version: Arc<PinnedVersion>,
+    pub new_pinned_version: PinnedVersion,
+}
+
+#[derive(Clone)]
+struct CacheRefillContext {
+    config: Arc<CacheRefillConfig>,
+    concurrency: Arc<Semaphore>,
+    sstable_store: SstableStoreRef,
+}
+
+pub struct CacheRefillTask {
+    deltas: Vec<SstDeltaInfo>,
+    context: CacheRefillContext,
+}
+
+impl CacheRefillTask {
+    async fn run(self) {
+        let tasks = self
+            .deltas
+            .iter()
+            .map(|delta| {
+                let context = self.context.clone();
+                async move {
+                    let holders = match Self::meta_cache_refill(&context, delta).await {
+                        Ok(holders) => holders,
+                        Err(e) => {
+                            tracing::warn!("meeta cache refill error: {:?}", e);
```

Comment:
nits: `meeta` -> `meta`.

Should this be an error log btw?

Reply:
> Should this be an error log btw?

IMO, an error in cache refill can be ignored in most cases.